### ISSUE TITLE
Nested objects with inner hits query returns "extracted source isn't an object or an array"

### ISSUE THREAD

**Issue by convell** [strapdata/elassandra#346]
**Elassandra version**: Behavior is seen on official docker images [6.8.4.0](https://hub.docker.com/layers/strapdata/elassandra/6.8.4.0/images/sha256-1e004ecccfc506faf806464e4beed1815d3c973ee6d3bb4b88193084397d8491?context=explore) -> [6.8.4.5](https://hub.docker.com/layers/strapdata/elassandra/6.8.4.5/images/sha256-1a7ff3e3c1ee0801f812aa321dfe4ae8a6f959c3ef47b0fbeaa3a4de695eff14?context=explore)

**Plugins installed**: None

**JVM version** (`java -version`): Docker image of 6.8.4.5 is running openjdk version "1.8.0_252"

**OS version** (`uname -a` if on a Unix-like system): Running on a container on a host that is running 18.04.1-Ubuntu

**Description of the problem including expected versus actual behavior**: 
I am moving from 6.2.3.28 to 6.8.4.5, and have noticed queries with inner hits on nested objects (as outlined in the steps below) are failing with the following error coming back: 
```
{
    "error": {
        "root_cause": [
            {
                "type": "illegal_state_exception",
                "reason": "extracted source isn't an object or an array"
            }
        ],
        "type": "search_phase_execution_exception",
        "reason": "all shards failed",
        "phase": "query",
        "grouped": true,
        "failed_shards": [
            {
                "shard": 0,
                "index": "nested_data",
                "node": "5c622bfa-fef0-4dd0-b040-5a68a767beb5",
                "reason": {
                    "type": "illegal_state_exception",
                    "reason": "extracted source isn't an object or an array"
                }
            }
        ],
        "caused_by": {
            "type": "illegal_state_exception",
            "reason": "extracted source isn't an object or an array",
            "caused_by": {
                "type": "illegal_state_exception",
                "reason": "extracted source isn't an object or an array"
            }
        }
    },
    "status": 500
}
```
**Reasons why I think this is a bug:** 
1. When tested against [ES 6.8.4 docker image](https://hub.docker.com/layers/elasticsearch/library/elasticsearch/6.8.4/images/sha256-87538ba78df48470563796df1dbbf2e7bf407e97542830bfd395f912b15c07f0?context=explore) I do not see the same issue. 
2. I do not see this issue on Elassandra 6.2.3.28, and assume there is backwards compatibility of query/mapping/docs on non major version releases 

**Other behaviors I noted:**
Removing the inner_hits block of the query stops the issue, and so does turning off the source in the inner_hits block. However neither of these are desirable solutions

Something else that seems to resolve the issue while I was triaging is removing `nested-object2` when inserting the data in step 2 below

**Steps to reproduce**:

 1. Create index at `nested_data` with the following mapping:
```
{
  "mappings": {
    "nested_data": {
      "properties": {
        "parent-nested": {
          "type": "nested",
          "properties": {
            "nested-object1": {
              "type": "nested",
              "properties": {
                "field1": {
                  "type": "text"
                },
                "field2": {
                  "type": "text",
                  "fields": {
                    "keyword": {
                      "type": "keyword",
                      "ignore_above": 256
                    }
                  }
                }
              }
            },
            "nested-object2": {
              "type": "nested",
              "properties": {
                "field2": {
                  "type": "text"
                }
              }
            }
          }
        }
      }
    }
  }
}
``` 
 2. Insert the following document in the above created index:
```
{
  "parent-nested": [
    {
      "nested-object1": [
        {
          "field1": "hello"
        }
      ]
    },
    {
      "nested-object2": {
      }
    }
  ]
}
```
 3. Run the following query against the index:
```
{
  "query": {
    "bool": {
      "must": [
        {
          "nested": {
            "path": "parent-nested.nested-object1",
            "inner_hits": {
            },
            "query": {
              "bool": {
                "should": {
                  "bool": {
                    "must": {
                      "match": {
                        "parent-nested.nested-object1.field1": "hello"
                      }
                    }
                  }
                }
              }
            }
          }
        }
      ]
    }
  }
}
```

**Please provide the following information**:

* elassandra logs:
```
elassandra           | Elassandra started
elassandra           | 2020-06-04 17:51:31,748 WARN  [elasticsearch[172.18.0.2][http_server_worker][T#2]] org.elasticsearch.common.logging.DeprecationLogger.deprecated(DeprecationLogger.java:242) [types removal] The parameter include_type_name should be explicitly specified in create index requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', and requests are expected to omit the type name in mapping definitions.
elassandra           | 2020-06-04 17:51:33,174 INFO  [elasticsearch[172.18.0.2][generic][T#2]] org.elassandra.shard.CassandraShardStateListener.afterIndexShardStarted(CassandraShardStateListener.java:69) shard [nested_data][0] started
elassandra           | 2020-06-04 17:51:46,971 WARN  [elasticsearch[172.18.0.2][search][T#1]] org.elasticsearch.rest.BytesRestResponse.build(BytesRestResponse.java:133) path: /nested_data/_search, params: {index=nested_data}
elassandra           | org.elasticsearch.action.search.SearchPhaseExecutionException: all shards failed
elassandra           |  at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:298)
elassandra           |  at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:135)
elassandra           |  at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:261)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase.onShardFailure(InitialSearchPhase.java:100)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase.access$100(InitialSearchPhase.java:48)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase$2.lambda$onFailure$1(InitialSearchPhase.java:220)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase.maybeFork(InitialSearchPhase.java:174)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase.access$000(InitialSearchPhase.java:48)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase$2.onFailure(InitialSearchPhase.java:220)
elassandra           |  at org.elasticsearch.action.search.SearchExecutionStatsCollector.onFailure(SearchExecutionStatsCollector.java:73)
elassandra           |  at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:59)
elassandra           |  at org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:463)
elassandra           |  at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1118)
elassandra           |  at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1230)
elassandra           |  at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1204)
elassandra           |  at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:60)
elassandra           |  at org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:56)
elassandra           |  at org.elasticsearch.search.SearchService$2.onFailure(SearchService.java:368)
elassandra           |  at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:362)
elassandra           |  at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:356)
elassandra           |  at org.elasticsearch.search.SearchService$4.doRun(SearchService.java:1114)
elassandra           |  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
elassandra           |  at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41)
elassandra           |  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:751)
elassandra           |  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
elassandra           |  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
elassandra           |  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
elassandra           |  at java.lang.Thread.run(Thread.java:748)
elassandra           | Caused by: org.elasticsearch.ElasticsearchException$1: extracted source isn't an object or an array
elassandra           |  at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:657)
elassandra           |  at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:133)
elassandra           |  ... 26 common frames omitted
elassandra           | Caused by: java.lang.IllegalStateException: extracted source isn't an object or an array
elassandra           |  at org.elasticsearch.search.fetch.FetchPhase.createNestedSearchHit(FetchPhase.java:340)
elassandra           |  at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:166)
elassandra           |  at org.elasticsearch.search.fetch.subphase.InnerHitsFetchSubPhase.hitsExecute(InnerHitsFetchSubPhase.java:69)
elassandra           |  at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:184)
elassandra           |  at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:424)
elassandra           |  at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:404)
elassandra           |  at org.elasticsearch.search.SearchService.access$100(SearchService.java:127)
elassandra           |  at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:360)
elassandra           |  ... 9 common frames omitted
elassandra           | 2020-06-04 18:06:06,393 WARN  [elasticsearch[172.18.0.2][search][T#2]] org.elasticsearch.rest.BytesRestResponse.build(BytesRestResponse.java:133) path: /nested_data/_search, params: {index=nested_data}
elassandra           | org.elasticsearch.action.search.SearchPhaseExecutionException: all shards failed
elassandra           |  at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:298)
elassandra           |  at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:135)
elassandra           |  at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:261)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase.onShardFailure(InitialSearchPhase.java:100)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase.access$100(InitialSearchPhase.java:48)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase$2.lambda$onFailure$1(InitialSearchPhase.java:220)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase.maybeFork(InitialSearchPhase.java:174)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase.access$000(InitialSearchPhase.java:48)
elassandra           |  at org.elasticsearch.action.search.InitialSearchPhase$2.onFailure(InitialSearchPhase.java:220)
elassandra           |  at org.elasticsearch.action.search.SearchExecutionStatsCollector.onFailure(SearchExecutionStatsCollector.java:73)
elassandra           |  at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:59)
elassandra           |  at org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:463)
elassandra           |  at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1118)
elassandra           |  at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1230)
elassandra           |  at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1204)
elassandra           |  at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:60)
elassandra           |  at org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:56)
elassandra           |  at org.elasticsearch.search.SearchService$2.onFailure(SearchService.java:368)
elassandra           |  at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:362)
elassandra           |  at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:356)
elassandra           |  at org.elasticsearch.search.SearchService$4.doRun(SearchService.java:1114)
elassandra           |  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
elassandra           |  at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41)
elassandra           |  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:751)
elassandra           |  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
elassandra           |  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
elassandra           |  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
elassandra           |  at java.lang.Thread.run(Thread.java:748)
elassandra           | Caused by: org.elasticsearch.ElasticsearchException$1: extracted source isn't an object or an array
elassandra           |  at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:657)
elassandra           |  at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:133)
elassandra           |  ... 26 common frames omitted
elassandra           | Caused by: java.lang.IllegalStateException: extracted source isn't an object or an array
elassandra           |  at org.elasticsearch.search.fetch.FetchPhase.createNestedSearchHit(FetchPhase.java:340)
elassandra           |  at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:166)
elassandra           |  at org.elasticsearch.search.fetch.subphase.InnerHitsFetchSubPhase.hitsExecute(InnerHitsFetchSubPhase.java:69)
elassandra           |  at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:184)
elassandra           |  at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:424)
elassandra           |  at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:404)
elassandra           |  at org.elasticsearch.search.SearchService.access$100(SearchService.java:127)
elassandra           |  at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:360)
elassandra           |  ... 9 common frames omitted
```

* cassandra schema: 
```
CREATE KEYSPACE nested_data WITH replication = {'class': 'NetworkTopologyStrategy', 'DC1': '1'}  AND durable_writes = true;

CREATE TYPE nested_data.nested_data_parent_nested_nested_object1 (
    field1 list<text>,
    field2 list<text>
);

CREATE TYPE nested_data.nested_data_parent_nested_nested_object2 (
    field2 list<text>
);

CREATE TYPE nested_data.nested_data_parent_nested (
    "nested-object1" list<frozen<nested_data_parent_nested_nested_object1>>,
    "nested-object2" list<frozen<nested_data_parent_nested_nested_object2>>
);

CREATE TABLE nested_data.nested_data (
    "_id" text PRIMARY KEY,
    "parent-nested" list<frozen<nested_data_parent_nested>>
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';
CREATE CUSTOM INDEX elastic_nested_data_idx ON nested_data.nested_data () USING 'org.elassandra.index.ExtendedElasticSecondaryIndex';

```

**Comment by vroyer** [comment]
Yes, this is a bug in elassandra version 6.8.4.x. 
A change in the elasticseach 6.5 internal storage of nested documents was not properly implemented in elassandra 6.8.4.x. This is fixed for the next release.
Thank's for reporting a complete reproductible test case .

**Comment by convell** [comment]
Thank you for the quick response, I look forward to the next release!

**Comment by vroyer** [comment]
Fixed in release 6.8.4.6
